### PR TITLE
Mount local artifact folder to docker container in MLProject

### PR DESCRIPTION
**Pull request by nlaille**
## What changes are proposed in this pull request?
 
Fix issue : #1450 

Mount artifact folder to docker container if artifact_uri is local.
 
## How is this patch tested?
 
Add a new unit test in `tests/projects/test_docker_projects.py`

Manual test with `examples/docker`
 
### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
Fix issue on logging artifact using local file system as artifact storage and docker MLProject.
 
### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [ ] Tracking
- [x] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by mateiz** [comment]
This is a good catch but unfortunately it doesn't merge cleanly anymore. Do you mind updating it to fix the merge conflicts? Hopefully it won't need a large change. Thanks for adding a test too!

**Comment by nlaille** [comment]
@mateiz done. 
Btw an extension to this one : #1621

**Comment by smurching** [comment]
Thanks @nlaille for the PR! Just to make sure things work end-to-end, would you mind:
1) Adding logic to [this example docker project script](https://github.com/mlflow/mlflow/blob/fbe8759ebe259bf8e0ca165f21d90be08b6a00f9/tests/resources/example_docker_project/scripts/docker_tracking_test.py#L7) that logs an artifact, and 
2) Updating [this test](https://github.com/mlflow/mlflow/blob/fbe8759ebe259bf8e0ca165f21d90be08b6a00f9/tests/projects/test_docker_projects.py#L56) to verify that the artifact was indeed logged? You should be able to use the MlflowClient's list_artifacts API ([see docs](https://mlflow.org/docs/latest/python_api/mlflow.tracking.html#mlflow.tracking.MlflowClient.list_artifacts), an MLflowClient object is already available in the test as `mlflow_service`) to verify that the artifact is present.

**Comment by smurching** [comment]
To clarify, the PR otherwise LGTM - thanks!

**Comment by nlaille** [comment]
@smurching I added a way to handle relative path as well. Luckily the test highlighted it :)

File: mlflow/projects/__init__.py [hunk 1]
```diff
@@ -695,9 +696,17 @@ def _get_docker_command(image, active_run):
...
     tracking_uri = tracking.get_tracking_uri()
     local_path, container_tracking_uri = _get_local_uri_or_none(tracking_uri)
+    artifact_uri_local_path = get_local_path_or_none(active_run.info.artifact_uri)
     if local_path is not None:
         cmd += ["-v", "%s:%s" % (local_path, _MLFLOW_DOCKER_TRACKING_DIR_PATH)]
         env_vars[tracking._TRACKING_URI_ENV_VAR] = container_tracking_uri
+    if artifact_uri_local_path is not None:
+        container_path = artifact_uri_local_path
+        if not os.path.isabs(container_path):
+            container_path = os.path.join("/mlflow/projects/code/", artifact_uri_local_path)
```
Comment (smurching) [inline mlflow/projects/__init__.py]: Would be nice to pull "/mlflow/projects/code/" into a constant so we don't have this string duplicated across `mlflow/projects/__init__.py`, if you want to make a follow-up PR to do that :)

**Review by smurching — APPROVED** [review]
LGTM thanks @nlaille for the updates!

**Comment by leobenkel** [comment]
Hello,
I am still having an issue on this topic.
I have one container that hold MLFLow server.
And one container that hold Jupyter notebook.
When the container #2 runs `mlflow.sklearn.log_model`, the model is being saved locally under `/tmp/mlflow` instead of in the container #1. 

Both of them have a mounted volume for the artifact folder but I can't find how to set up a default path in the client.

**Comment by sucitw** [comment]
@leobenkel,

I had the same issue to access artifacts when running MLflow server via container.

According to this [stackoverflow](https://stackoverflow.com/questions/52331254/how-to-store-artifacts-on-a-server-running-mlflow) 

> The best practice for artifact storage with a remote tracking server is to configure the server to use an artifact root accessible to both clients and the server.

Therefore, for the container approach, you need to use docker volume to mount the artifact folder for each container. Then, you are supposed to see artifact from the client and server-side.

**Comment by leobenkel** [comment]
Thank you @sucitw !

**Pull request merged**